sw/device/lib/base/hardened_memory.c

```diff
@@ -0,0 +1,180 @@
+
+  size_t count = 0;
+  size_t expected_count = random_order_len(&order);
+
+  uintptr_t data_addr = (uintptr_t)dest;
+
+  uint32_t decoys[8];
+  uintptr_t decoy_addr = (uintptr_t)&decoys;
+
+  for (; count < expected_count; count = launderw(count) + 1) {
+    size_t idx = launderw(random_order_advance(&order)) * 4;
+    barrierw(idx);
+
+    uintptr_t datap = data_addr + idx;
+    uintptr_t decoy = decoy_addr + (idx & 7);
+
+    void *data = (void *)launderw(
+        ct_cmovw(ct_sltuw(launderw(idx), word_len), datap, decoy));
+
+    // Write a freshly-generated random word to `*data`.
+    write_32(hardened_memshred_random_word(), data);
+  }
+
+  HARDENED_CHECK_EQ(count, expected_count);
+}
+
+hardened_bool_t hardened_memeq(const uint32_t *lhs, const uint32_t *rhs,
+                               size_t word_len) {
+  random_order_t order;
+  random_order_init(&order, word_len);
+
+  size_t count = 0;
+  size_t expected_count = random_order_len(&order);
+
+  uintptr_t a_addr = (uintptr_t)lhs;
+  uintptr_t b_addr = (uintptr_t)rhs;
+
+  // `decoys` needs to be filled with equal values this time around. It
+  // should be filled with values with a Hamming weight of around 16, which is
+  // the most common hamming weight among 32-bit words.
+  uint32_t decoys[8] = {
+      0xaaaaaaaa, 0xaaaaaaaa, 0xaaaaaaaa, 0xaaaaaaaa,
+      0xaaaaaaaa, 0xaaaaaaaa, 0xaaaaaaaa, 0xaaaaaaaa,
+  };
+  uintptr_t decoy_addr = (uintptr_t)&decoys;
+
+  uint32_t zeros = 0;
+  uint32_t ones = UINT32_MAX;
+
+  // The loop is almost token-for-token the one above, but the copy is
+  // replaced with something else.
+  for (; count < expected_count; count = launderw(count) + 1) {
+    size_t idx = launderw(random_order_advance(&order)) * 4;
+    barrierw(idx);
+
+    uintptr_t ap = a_addr + idx;
+    uintptr_t bp = b_addr + idx;
+    uintptr_t decoy1 = decoy_addr + (idx & 7);
+    uintptr_t decoy2 = decoy_addr + ((idx + 4) & 7);
+
+    void *av = (void *)launderw(
+        ct_cmovw(ct_sltuw(launderw(idx), word_len), ap, decoy1));
+    void *bv = (void *)launderw(
+        ct_cmovw(ct_sltuw(launderw(idx), word_len), bp, decoy2));
+
+    uint32_t a = read_32(av);
+    uint32_t b = read_32(bv);
+
+    // Launder one of the operands, so that the compiler cannot cache the result
+    // of the xor for use in the next operation.
+    //
+    // We launder `zeroes` so that compiler cannot learn that `zeroes` has
+    // strictly more bits set at the end of the loop.
+    zeros = launder32(zeros) | (launder32(a) ^ b);
+
+    // Same as above. The compiler can cache the value of `a[offset]`, but it
+    // has no chance to strength-reduce this operation.
+    ones = launder32(ones) & (launder32(a) ^ ~b);
+  }
+
+  // NOTE(moidx): shuffled_memcmp from #6924 has a better method to ensure that
+  // all the addresses where traversed.
+
+  HARDENED_CHECK_EQ(count, expected_count);
+  if (launder32(zeros) == 0) {
+    HARDENED_CHECK_EQ(ones, UINT32_MAX);
```

Comment:
should we have a HARDENED_CHECK here for zeros after ones here and below?

Reply:
Not actually necessary; the purpose of checking for ones is that it's *stronger* than checking for zeros, since it's a distinct value that needs to be attacked. Checking for zero adds more code for minimal value, IMO.